File: src/spatial_operators/vorticities.f90
module vorticities

  ! Here, vorticities are calculated. The word "vorticity" hereby refers to both vertical and tangential components.

  use iso_c_binding
  use definitions, only: wp
  use grid_nml,    only: n_layers,n_vectors_h,n_vectors,n_layers,n_dual_vectors_per_layer,n_dual_v_vectors, &
                         n_dual_scalars_h,n_scalars_h,n_vectors_per_layer,n_vectors_h,n_oro_layers,n_dual_vectors, &
                         radius
  
  implicit none
  
  contains

  subroutine calc_rel_vort_on_triangles(velocity_field,out_field,vorticity_indices_triangles,area_dual, &
                                        z_vector,z_vector_dual,vorticity_signs_triangles,normal_distance) &
  bind(c,name = "calc_rel_vort_on_triangles")
    
    ! This subroutine calculates the vertical relative vorticity on triangles.
    
    real(wp), intent(in)  :: velocity_field(n_vectors),z_vector(n_vectors),z_vector_dual(n_dual_vectors), &
                             normal_distance(n_vectors),area_dual(n_dual_vectors)
    integer,  intent(in)  :: vorticity_indices_triangles(3*n_dual_scalars_h), &
                             vorticity_signs_triangles(3*n_dual_scalars_h)
    real(wp), intent(out) :: out_field(n_dual_v_vectors)
    
    ! local variables
    integer  :: ji,jk,layer_index,h_index,vector_index,index_for_vertical_gradient
    real(wp) :: velocity_value,length_rescale_factor,vertical_gradient,delta_z
    
    ! loop over all triangles
    !$omp parallel do private(ji,jk,layer_index,h_index,velocity_value,length_rescale_factor, &
    !$omp vector_index,index_for_vertical_gradient,vertical_gradient,delta_z)
    do ji=1,n_dual_v_vectors
      layer_index = (ji-1)/n_dual_scalars_h
      h_index = ji - layer_index*n_dual_scalars_h
      ! clearing what has previously been here
      out_field(ji) = 0._wp
      ! loop over the three edges of the triangle at hand
      do jk=1,3
        vector_index = n_scalars_h + layer_index*n_vectors_per_layer + 1+vorticity_indices_triangles(3*(h_index-1)+jk)
        velocity_value = velocity_field(vector_index)
        ! this corrects for terrain following coordinates
        length_rescale_factor = 1._wp
          if (layer_index>=n_layers-n_oro_layers) then
            length_rescale_factor = (radius + z_vector_dual(n_vectors_h + layer_index*n_dual_vectors_per_layer + h_index)) &
            /(radius + z_vector(vector_index))
            delta_z = z_vector_dual(n_vectors_h + layer_index*n_dual_vectors_per_layer + h_index) - z_vector(vector_index)
            if (delta_z>0._wp) then
              index_for_vertical_gradient = vector_index - n_vectors_per_layer
            else
              if (layer_index==n_layers-1) then
                index_for_vertical_gradient = vector_index - n_vectors_per_layer
              else
                index_for_vertical_gradient = vector_index + n_vectors_per_layer
              endif
            endif
            vertical_gradient = (velocity_field(vector_index) - velocity_field(index_for_vertical_gradient)) &
                                /(z_vector(vector_index) - z_vector(index_for_vertical_gradient))
            ! Here, the vertical interpolation is made.
            velocity_value = velocity_value+delta_z*vertical_gradient
        endif
        out_field(ji) = out_field(ji) + length_rescale_factor*normal_distance(vector_index) &
                                        *vorticity_signs_triangles(3*(h_index-1)+jk)*velocity_value
      enddo
      ! dividing by the area (Stokes' Theorem)
      out_field(ji) = out_field(ji)/area_dual(n_vectors_h + layer_index*n_dual_vectors_per_layer + h_index)
    
    enddo
    !$omp end parallel do
    
  end subroutine calc_rel_vort_on_triangles

  subroutine add_f_to_rel_vort(rel_vort,f_vec,out_field) &
  bind(c,name = "add_f_to_rel_vort")
  
    ! This subroutine adds the Coriolis parameter to the relative vorticity.
    
    real(wp), intent(in)  :: rel_vort(n_layers*2*n_vectors_h+n_vectors_h),f_vec(2*n_vectors_h)
    real(wp), intent(out) :: out_field(n_layers*2*n_vectors_h+n_vectors_h)
    
    integer :: ji,layer_index,h_index
    
    !$omp parallel do private(ji,layer_index,h_index)
    do ji=1,n_layers*2*n_vectors_h+n_vectors_h
      layer_index = (ji-1)/(2*n_vectors_h)
      h_index = ji - layer_index*2*n_vectors_h
      out_field(ji) = rel_vort(ji) + f_vec(h_index)
    enddo
    !$omp end parallel do
  
  end subroutine add_f_to_rel_vort
  
end module vorticities

File: src/spatial_operators/vorticities.c
/*
This source file is part of the Geophysical Fluids Modeling Framework (GAME), which is released under the MIT license.
Github repository: https://github.com/OpenNWP/GAME
*/

/*
Here, vorticities are calculated. The word "vorticity" hereby refers to both vertical and tangential components.
*/

#include <stdio.h>
#include "../game_types.h"
#include "../constituents/constituents.h"
#include "spatial_operators.h"

extern int add_f_to_rel_vort();
extern int calc_rel_vort_on_triangles();
extern int calc_rel_vort();

int calc_pot_vort(Vector_field velocity_field, Scalar_field density_field, Diagnostics *diagnostics, Grid *grid, Dualgrid *dualgrid)
{
	// It is called "potential vorticity", but it is not Ertel's potential vorticity. It is the absolute vorticity divided by the density.
	calc_rel_vort(velocity_field,diagnostics->rel_vort_on_triangles,grid->area,grid->z_vector,dualgrid->z_vector,diagnostics->rel_vort, &
                  dualgrid->vorticity_indices_triangles,dualgrid->vorticity_signs_triangles,grid->normal_distance, &
                  dualgrid->area,grid->from_index,grid->to_index,dualgrid->from_index,dualgrid->to_index,grid->inner_product_weights, &
                  grid->slope);
	// pot_vort is a misuse of name here
	add_f_to_rel_vort(diagnostics -> rel_vort, dualgrid -> f_vec, diagnostics -> pot_vort);
    int layer_index, h_index, edge_vector_index_h, upper_from_index, upper_to_index;
    double density_value;
    // determining the density value by which we need to divide
    #pragma omp parallel for private (layer_index, h_index, edge_vector_index_h, upper_from_index, upper_to_index, density_value)
    for (int i = 0; i < N_LAYERS*2*N_VECS_H + N_VECS_H; ++i)
    {
        layer_index = i/(2*N_VECS_H);
        h_index = i - layer_index*2*N_VECS_H;
        // interpolation of the density to the center of the rhombus
        if (h_index >= N_VECS_H)
        {
			edge_vector_index_h = h_index - N_VECS_H;
			density_value = 0;
			for (int j = 0; j < 4; ++j)
			{
				density_value
				+= grid -> density_to_rhombi_weights[4*edge_vector_index_h + j]
				*density_field[layer_index*N_SCALS_H + grid -> density_to_rhombi_indices[4*edge_vector_index_h + j]];
			}
        }
        // interpolation of the density to the half level edges
        else
        {
        	// linear extrapolation to the TOA
        	if (layer_index == 0)
        	{
				density_value
				= 0.5*(density_field[grid -> from_index[h_index]] + density_field[grid -> to_index[h_index]])
				// the gradient
				+ (0.5*(density_field[grid -> from_index[h_index]] + density_field[grid -> to_index[h_index]])
				- 0.5*(density_field[grid -> from_index[h_index] + N_SCALS_H] + density_field[grid -> to_index[h_index] + N_SCALS_H]))
				/(grid -> z_vector[N_SCALARS + h_index] - grid -> z_vector[N_SCALARS + N_VECS_PER_LAYER + h_index])
				// delta z
				*(grid -> z_vector[0] - grid -> z_vector[N_SCALARS + h_index]);
        	}
        	// linear extrapolation to the surface
            else if (layer_index == N_LAYERS)
            {
				density_value =
				0.5*(density_field[(layer_index - 1)*N_SCALS_H + grid -> from_index[h_index]] + density_field[(layer_index - 1)*N_SCALS_H + grid -> to_index[h_index]])
				// the gradient
				+ (0.5*(density_field[(layer_index - 2)*N_SCALS_H + grid -> from_index[h_index]] + density_field[(layer_index - 2)*N_SCALS_H + grid -> to_index[h_index]])
				- 0.5*(density_field[(layer_index - 1)*N_SCALS_H + grid -> from_index[h_index]] + density_field[(layer_index - 1)*N_SCALS_H + grid -> to_index[h_index]]))
				/(grid -> z_vector[N_SCALARS + (layer_index - 2)*N_VECS_PER_LAYER + h_index] - grid -> z_vector[N_SCALARS + (layer_index - 1)*N_VECS_PER_LAYER + h_index])
				// delta z
				*(0.5*(grid -> z_vector[layer_index*N_VECS_PER_LAYER + grid -> from_index[h_index]] + grid -> z_vector[layer_index*N_VECS_PER_LAYER + grid -> to_index[h_index]])
				- grid -> z_vector[N_SCALARS + (layer_index - 1)*N_VECS_PER_LAYER + h_index]);
            }
            else
            {
            	upper_from_index = (layer_index - 1)*N_SCALS_H + grid -> from_index[h_index];
            	upper_to_index = (layer_index - 1)*N_SCALS_H + grid -> to_index[h_index];
            	density_value = 0.25*(density_field[upper_from_index] + density_field[upper_to_index]
            	+ density_field[upper_from_index + N_SCALS_H] + density_field[upper_to_index + N_SCALS_H]);
            }
        }
        
        // division by the density to obtain the "potential vorticity"
		diagnostics -> pot_vort[i] = diagnostics -> pot_vort[i]/density_value;
    }
    return 0;
}













